# Imports

In [68]:
from classes.FFNN import FFNN
from sklearn import datasets
import numpy as np
from algorithm.Util import save

# Load Datasets Iris

In [69]:
iris = datasets.load_iris()
x = iris.data
y = []
for test in iris.target : 
    temp = []
    temp.append(test)
    y.append(temp)
print(len(y))

150


In [70]:
print(len(x))

150


In [71]:
#data split
x_train = x[0:int(len(x)*0.9)] 
x_test = x[int(len(x)*0.9):]

y_train = y[0:int(len(x)*0.9)] 
y_test = y[int(len(x)*0.9):]



# Load Model

## Iris Model

In [72]:
#iris model
Model_Test = FFNN('./file/models/Iris_Model_Test.json')
Model_Test.setLearningRate(0.01)

NOTICE: MODEL Iris LOADED SUCCESSFULLY!


# Train model

In [73]:
#iris model
#train using all of model
Model_Test.train(x, y, 200, 1, 0.0000000000001)


Epoch 0 / 200
1/1 [==============================] loss:0
Epoch 1 / 200
1/1 [==============================] loss:0
Epoch 2 / 200
1/1 [==============================] loss:0
Epoch 3 / 200
1/1 [==============================] loss:0
Epoch 4 / 200
1/1 [==============================] loss:0
Epoch 5 / 200
1/1 [==============================] loss:0
Epoch 6 / 200
1/1 [==============================] loss:0
Epoch 7 / 200
1/1 [==============================] loss:0
Epoch 8 / 200
1/1 [==============================] loss:0
Epoch 9 / 200
1/1 [==============================] loss:0
Epoch 10 / 200
1/1 [==============================] loss:0
Epoch 11 / 200
1/1 [==============================] loss:0
Epoch 12 / 200
1/1 [==============================] loss:0
Epoch 13 / 200
1/1 [==============================] loss:0
Epoch 14 / 200
1/1 [==============================] loss:0
Epoch 15 / 200
1/1 [==============================] loss:0
Epoch 16 / 200
1/1 [==============================] loss:0
Epoch 

## Batch Prediction

In [74]:
batch = x

In [75]:
mymodel_predict = Model_Test.batch_predict(batch)
print(f'Batch prediction = \n')

#reformat answers
finalAns = []
for item in mymodel_predict:
    if item[0]>0.5:
        finalAns.append(1)
    else:
        finalAns.append(0)
print(finalAns)



Batch prediction = 

[1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]


In [76]:
from sklearn.neural_network import MLPClassifier

clf = MLPClassifier(random_state=1, max_iter=100)
clf.fit(x,y)

mlp_predict = clf.predict(x)
print(mlp_predict)

[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1 1 2 1 2 2 1 1 2 1 1 1 2 1 1 2 1 2 1 2 1 2 2
 1 1 1 2 2 1 1 1 1 2 2 2 1 1 2 2 2 2 1 1 2 2 2 1 1 1 2 2 2 2 2 2 2 2 2 2 2
 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2
 2 2]


C:\Users\hp\AppData\Local\Programs\Python\Python310\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:1109: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
C:\Users\hp\AppData\Local\Programs\Python\Python310\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:692: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (100) reached and the optimization hasn't converged yet.
  warnings.warn(


In [77]:
import algorithm.Util

In [78]:
#kinerja sklearn 
confusionMatrix = algorithm.Util.findConfusionMatrix(iris.target, mlp_predict)
print("confusion matrix: ", end="\n")
print(confusionMatrix)
print("we have multiple classified data = setosa, versicolor, virginica : 0,1,2 <- in that order")
print("accuracy: " + str(algorithm.Util.getAccuracy(confusionMatrix)))
print("macro precision: " + str(algorithm.Util.getMacroPrecision(confusionMatrix)))
print("macro recall: " + str(algorithm.Util.getMacroRecall(confusionMatrix)))
print("macro f1 score: " + str(algorithm.Util.getMacroF1Score(confusionMatrix)))

confusion matrix: 
[[50, 0, 0], [0, 28, 22], [0, 0, 50]]
we have multiple classified data = setosa, versicolor, virginica : 0,1,2 <- in that order
accuracy: 0.8533333333333334
macro precision: 0.8981481481481483
macro recall: 0.8533333333333334
macro f1 score: 0.8751674067808557


In [79]:
#kinerja model
confusionMatrix = algorithm.Util.findConfusionMatrix(iris.target, finalAns)
print("confusion matrix: ", end="\n")
print(confusionMatrix)
print("we have multiple classified data = setosa, versicolor, virginica : 0,1,2 <- in that order")
print("accuracy: " + str(algorithm.Util.getAccuracy(confusionMatrix)))
print("macro precision: " + str(algorithm.Util.getMacroPrecision(confusionMatrix)))
print("macro recall: " + str(algorithm.Util.getMacroRecall(confusionMatrix)))
print("macro f1 score: " + str(algorithm.Util.getMacroF1Score(confusionMatrix)))

confusion matrix: 
[[0, 50, 0], [0, 50, 0], [0, 50, 0]]
we have multiple classified data = setosa, versicolor, virginica : 0,1,2 <- in that order
accuracy: 0.3333333333333333
macro precision: 0.1111111111111111
macro recall: 0.3333333333333333
macro f1 score: 0.16666666666666666


# Training Using Split Data

In [80]:
#redefine model
Model_Test = FFNN('./file/models/Iris_Model_Test.json')
Model_Test.setLearningRate(0.01)
#train using training data
Model_Test.train(x_train, y_train, 200, 1, 0.0000000000001)

NOTICE: MODEL Iris LOADED SUCCESSFULLY!

Epoch 0 / 200
1/1 [==============================] loss:0
Epoch 1 / 200
1/1 [==============================] loss:0
Epoch 2 / 200
1/1 [==============================] loss:0
Epoch 3 / 200
1/1 [==============================] loss:0
Epoch 4 / 200
1/1 [==============================] loss:0
Epoch 5 / 200
1/1 [==============================] loss:0
Epoch 6 / 200
1/1 [==============================] loss:0
Epoch 7 / 200
1/1 [==============================] loss:0
Epoch 8 / 200
1/1 [==============================] loss:0
Epoch 9 / 200
1/1 [==============================] loss:0
Epoch 10 / 200
1/1 [==============================] loss:0
Epoch 11 / 200
1/1 [==============================] loss:0
Epoch 12 / 200
1/1 [==============================] loss:0
Epoch 13 / 200
1/1 [==============================] loss:0
Epoch 14 / 200
1/1 [==============================] loss:0
Epoch 15 / 200
1/1 [==============================] loss:0
Epoch 16 / 200
1/1 [=====

In [81]:
#batch predict for test data
mymodel_predict = Model_Test.batch_predict(x_test)
print(f'Batch prediction = \n')

#reformat answers
finalAns = []
for item in mymodel_predict:
    if item[0]>0.5:
        finalAns.append(1)
    else:
        finalAns.append(0)
print(finalAns)

Batch prediction = 

[1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]


In [106]:
#kinerja utk training dengan split data
#reformat karena y_test = [[0], [1], [1]] -> [0,1,2]
y_final_test = [y_test[i][0] for i in range(len(y_test))]
print(y_final_test)
confusionMatrix = algorithm.Util.findConfusionMatrix(y_final_test, finalAns)
print("confusion matrix: ", end="\n")
print(confusionMatrix)
print("we have multiple classified data = setosa, versicolor, virginica : 0,1,2 <- in that order")
print("accuracy: " + str(algorithm.Util.getAccuracy(confusionMatrix)))
print("macro precision: " + str(algorithm.Util.getMacroPrecision(confusionMatrix)))
print("macro recall: " + str(algorithm.Util.getMacroRecall(confusionMatrix)))
print("macro f1 score: " + str(algorithm.Util.getMacroF1Score(confusionMatrix)))

[0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2]
confusion matrix: 
[[0, 2, 0], [0, 50, 0], [0, 50, 0]]
we have multiple classified data = setosa, versicolor, virginica : 0,1,2 <- in that order
accuracy: 0.49019607843137253
macro precision: 0.16339869281045752
macro recall: 0.3333333333333333
macro f1 score: 0.21929824561403508


# Train dengan 10-fold Cross Validation

In [83]:
from random import randrange

#randomly split data
split = randrange(len(x))

x_train = x[0:split] 
x_test = x[split:]

y_train = y[0:split] 
y_test = y[split:]

print(len(x_train))
print(len(x_test))
print(len(y_train))
print(len(y_test))

48
102
48
102


In [84]:
#because its 10 fold cross validation, we train 10 times using train = 0.9, validate = 0.1
#agar bisa dilihat lebih jelas utk semua konfigurasi, we train it 10 times
#its a bit bruteforce but yaudah lah ya

#config0
x_train0 = x_train[0:int(0.9*len(x_train))]
y_train0 = y_train[0:int(0.9*len(y_train))]
x_validate0 = x_train[int(0.9*len(x_train)):]
y_validate0 = y_train[int(0.9*len(y_train)):]

#config1
x_train1 = x_train[0:int(0.1*len(x_train))]
x_train1_join = x_train[int(0.2*len(x_train)):]
x_train1 = np.append(x_train1,x_train1_join, axis=0)
y_train1 = y_train[0:int(0.1*len(y_train))]
y_train1_join = y_train[int(0.2*len(y_train)):]
y_train1 = np.append(y_train1,y_train1_join, axis=0)
x_validate1 = x_train[int(0.1*len(x_train)):int(0.2*len(x_train))]
y_validate1 = y_train[int(0.1*len(y_train)):int(0.2*len(y_train))]

#config2
x_train2 = x_train[0:int(0.2*len(x_train))]
x_train2_join = x_train[int(0.3*len(x_train)):]
x_train2 = np.append(x_train2,x_train2_join, axis=0)
y_train2 = y_train[0:int(0.2*len(y_train))]
y_train2_join = y_train[int(0.3*len(y_train)):]
y_train2 = np.append(y_train2,y_train2_join, axis=0)
x_validate2 = x_train[int(0.2*len(x_train)):int(0.3*len(x_train))]
y_validate2 = y_train[int(0.2*len(y_train)):int(0.3*len(y_train))]

#config3
x_train3 = x_train[0:int(0.3*len(x_train))]
x_train3_join = x_train[int(0.4*len(x_train)):]
x_train3 = np.append(x_train3,x_train3_join, axis=0)
y_train3 = y_train[0:int(0.3*len(y_train))]
y_train3_join = y_train[int(0.4*len(y_train)):]
y_train3 = np.append(y_train3,y_train3_join, axis=0)
x_validate3 = x_train[int(0.3*len(x_train)):int(0.4*len(x_train))]
y_validate3 = y_train[int(0.3*len(y_train)):int(0.4*len(y_train))]

#config4
x_train4 = x_train[0:int(0.4*len(x_train))]
x_train4_join = x_train[int(0.5*len(x_train)):]
x_train4 = np.append(x_train4,x_train4_join, axis=0)
y_train4 = y_train[0:int(0.4*len(y_train))]
y_train4_join = y_train[int(0.5*len(y_train)):]
y_train4 = np.append(y_train4,y_train4_join, axis=0)
x_validate4 = x_train[int(0.4*len(x_train)):int(0.5*len(x_train))]
y_validate4 = y_train[int(0.4*len(y_train)):int(0.5*len(y_train))]

#config5
x_train5 = x_train[0:int(0.5*len(x_train))]
x_train5_join = x_train[int(0.6*len(x_train)):]
x_train5 = np.append(x_train5,x_train5_join, axis=0)
y_train5 = y_train[0:int(0.5*len(y_train))]
y_train5_join = y_train[int(0.6*len(y_train)):]
y_train5 = np.append(y_train5,y_train5_join, axis=0)
x_validate5 = x_train[int(0.5*len(x_train)):int(0.6*len(x_train))]
y_validate5 = y_train[int(0.5*len(y_train)):int(0.6*len(y_train))]

#config6
x_train6 = x_train[0:int(0.6*len(x_train))]
x_train6_join = x_train[int(0.7*len(x_train)):]
x_train6 = np.append(x_train6,x_train6_join, axis=0)
y_train6 = y_train[0:int(0.6*len(y_train))]
y_train6_join = y_train[int(0.7*len(y_train)):]
y_train6 = np.append(y_train6,y_train6_join, axis=0)
x_validate6 = x_train[int(0.6*len(x_train)):int(0.7*len(x_train))]
y_validate6 = y_train[int(0.6*len(y_train)):int(0.7*len(y_train))]

#config7
x_train7 = x_train[0:int(0.7*len(x_train))]
x_train7_join = x_train[int(0.8*len(x_train)):]
x_train7 = np.append(x_train7,x_train7_join, axis=0)
y_train7 = y_train[0:int(0.7*len(y_train))]
y_train7_join = y_train[int(0.8*len(y_train)):]
y_train7 = np.append(y_train7,y_train7_join, axis=0)
x_validate7 = x_train[int(0.7*len(x_train)):int(0.8*len(x_train))]
y_validate7 = y_train[int(0.7*len(y_train)):int(0.8*len(y_train))]

#config8
x_train8 = x_train[0:int(0.8*len(x_train))]
x_train8_join = x_train[int(0.9*len(x_train)):]
x_train8 = np.append(x_train8,x_train8_join, axis=0)
y_train8 = y_train[0:int(0.8*len(y_train))]
y_train8_join = y_train[int(0.9*len(y_train)):]
y_train8 = np.append(y_train8,y_train8_join, axis=0)
x_validate8 = x_train[int(0.8*len(x_train)):int(0.9*len(x_train))]
y_validate8 = y_train[int(0.8*len(y_train)):int(0.9*len(y_train))]

#config9
x_train9 = x_train[int(0.1*len(x_train)):]
y_train9 = y_train[int(0.1*len(x_train)):]
x_validate9 = x_train[0:int(0.1*len(x_train))]
y_validate9 = y_train[0:int(0.1*len(y_train))]

In [85]:
#config 0
#redefine model
Model_Test0 = FFNN('./file/models/Iris_Model_Test.json')
Model_Test0.setLearningRate(0.01)
#train using training data
Model_Test0.train(x_train0, y_train0, 200, 1, 0.0000000000001)
#predict validate
mymodel_predict = Model_Test0.batch_predict(x_validate0)
#reformat answers
finalAns = []
for item in mymodel_predict:
    if item[0]>0.5:
        finalAns.append(1)
    else:
        finalAns.append(0)
print(finalAns)
#reformat y_validate
y_validate = [y_validate0[i][0] for i in range (len(y_validate0))]
print(y_validate)
confusionMatrix = algorithm.Util.findConfusionMatrix(y_validate, finalAns)
print("confusion matrix: ", end="\n")
print(confusionMatrix)
print("we have multiple classified data = setosa, versicolor, virginica : 0,1,2 <- in that order")
print("accuracy: " + str(algorithm.Util.getAccuracy(confusionMatrix)))
print("macro precision: " + str(algorithm.Util.getMacroPrecision(confusionMatrix)))
print("macro recall: " + str(algorithm.Util.getMacroRecall(confusionMatrix)))
print("macro f1 score: " + str(algorithm.Util.getMacroF1Score(confusionMatrix)))

NOTICE: MODEL Iris LOADED SUCCESSFULLY!

Epoch 0 / 200
1/1 [==============================] loss:0
Epoch 1 / 200
1/1 [==============================] loss:0
Epoch 2 / 200
1/1 [==============================] loss:0
Epoch 3 / 200
1/1 [==============================] loss:0
Epoch 4 / 200
1/1 [==============================] loss:0
Epoch 5 / 200
1/1 [==============================] loss:0
Epoch 6 / 200
1/1 [==============================] loss:0
Epoch 7 / 200
1/1 [==============================] loss:0
Epoch 8 / 200
1/1 [==============================] loss:0
Epoch 9 / 200
1/1 [==============================] loss:0
Epoch 10 / 200
1/1 [==============================] loss:0
Epoch 11 / 200
1/1 [==============================] loss:0
Epoch 12 / 200
1/1 [==============================] loss:0
Epoch 13 / 200
1/1 [==============================] loss:0
Epoch 14 / 200
1/1 [==============================] loss:0
Epoch 15 / 200
1/1 [==============================] loss:0
Epoch 16 / 200
1/1 [=====

In [86]:
#config 1
#redefine model
Model_Test1 = FFNN('./file/models/Iris_Model_Test.json')
Model_Test1.setLearningRate(0.01)
#train using training data
Model_Test1.train(x_train1, y_train1, 200, 1, 0.0000000000001)
#predict validate
mymodel_predict = Model_Test1.batch_predict(x_validate1)
#reformat answers
finalAns = []
for item in mymodel_predict:
    if item[0]>0.5:
        finalAns.append(1)
    else:
        finalAns.append(0)
print(finalAns)
#reformat y_validate
y_validate = [y_validate1[i][0] for i in range (len(y_validate1))]
print(y_validate)
confusionMatrix = algorithm.Util.findConfusionMatrix(y_validate, finalAns)
print("confusion matrix: ", end="\n")
print(confusionMatrix)
print("we have multiple classified data = setosa, versicolor, virginica : 0,1,2 <- in that order")
print("accuracy: " + str(algorithm.Util.getAccuracy(confusionMatrix)))
print("macro precision: " + str(algorithm.Util.getMacroPrecision(confusionMatrix)))
print("macro recall: " + str(algorithm.Util.getMacroRecall(confusionMatrix)))
print("macro f1 score: " + str(algorithm.Util.getMacroF1Score(confusionMatrix)))

NOTICE: MODEL Iris LOADED SUCCESSFULLY!

Epoch 0 / 200
1/1 [==============================] loss:0
Epoch 1 / 200
1/1 [==============================] loss:0
Epoch 2 / 200
1/1 [==============================] loss:0
Epoch 3 / 200
1/1 [==============================] loss:0
Epoch 4 / 200
1/1 [==============================] loss:0
Epoch 5 / 200
1/1 [==============================] loss:0
Epoch 6 / 200
1/1 [==============================] loss:0
Epoch 7 / 200
1/1 [==============================] loss:0
Epoch 8 / 200
1/1 [==============================] loss:0
Epoch 9 / 200
1/1 [==============================] loss:0
Epoch 10 / 200
1/1 [==============================] loss:0
Epoch 11 / 200
1/1 [==============================] loss:0
Epoch 12 / 200
1/1 [==============================] loss:0
Epoch 13 / 200
1/1 [==============================] loss:0
Epoch 14 / 200
1/1 [==============================] loss:0
Epoch 15 / 200
1/1 [==============================] loss:0
Epoch 16 / 200
1/1 [=====

In [87]:
#config 2
#redefine model
Model_Test2 = FFNN('./file/models/Iris_Model_Test.json')
Model_Test2.setLearningRate(0.01)
#train using training data
Model_Test2.train(x_train2, y_train2, 200, 1, 0.0000000000001)
#predict validate
mymodel_predict = Model_Test2.batch_predict(x_validate2)
#reformat answers
finalAns = []
for item in mymodel_predict:
    if item[0]>0.5:
        finalAns.append(1)
    else:
        finalAns.append(0)
print(finalAns)
#reformat y_validate
y_validate = [y_validate2[i][0] for i in range (len(y_validate2))]
print(y_validate)
confusionMatrix = algorithm.Util.findConfusionMatrix(y_validate, finalAns)
print("confusion matrix: ", end="\n")
print(confusionMatrix)
print("we have multiple classified data = setosa, versicolor, virginica : 0,1,2 <- in that order")
print("accuracy: " + str(algorithm.Util.getAccuracy(confusionMatrix)))
print("macro precision: " + str(algorithm.Util.getMacroPrecision(confusionMatrix)))
print("macro recall: " + str(algorithm.Util.getMacroRecall(confusionMatrix)))
print("macro f1 score: " + str(algorithm.Util.getMacroF1Score(confusionMatrix)))

NOTICE: MODEL Iris LOADED SUCCESSFULLY!

Epoch 0 / 200
1/1 [==============================] loss:0
Epoch 1 / 200
1/1 [==============================] loss:0
Epoch 2 / 200
1/1 [==============================] loss:0
Epoch 3 / 200
1/1 [==============================] loss:0
Epoch 4 / 200
1/1 [==============================] loss:0
Epoch 5 / 200
1/1 [==============================] loss:0
Epoch 6 / 200
1/1 [==============================] loss:0
Epoch 7 / 200
1/1 [==============================] loss:0
Epoch 8 / 200
1/1 [==============================] loss:0
Epoch 9 / 200
1/1 [==============================] loss:0
Epoch 10 / 200
1/1 [==============================] loss:0
Epoch 11 / 200
1/1 [==============================] loss:0
Epoch 12 / 200
1/1 [==============================] loss:0
Epoch 13 / 200
1/1 [==============================] loss:0
Epoch 14 / 200
1/1 [==============================] loss:0
Epoch 15 / 200
1/1 [==============================] loss:0
Epoch 16 / 200
1/1 [=====

In [88]:
#config 3
#redefine model
Model_Test3 = FFNN('./file/models/Iris_Model_Test.json')
Model_Test3.setLearningRate(0.01)
#train using training data
Model_Test3.train(x_train3, y_train3, 200, 1, 0.0000000000001)
#predict validate
mymodel_predict = Model_Test3.batch_predict(x_validate3)
#reformat answers
finalAns = []
for item in mymodel_predict:
    if item[0]>0.5:
        finalAns.append(1)
    else:
        finalAns.append(0)
print(finalAns)
#reformat y_validate
y_validate = [y_validate3[i][0] for i in range (len(y_validate3))]
print(y_validate)
confusionMatrix = algorithm.Util.findConfusionMatrix(y_validate, finalAns)
print("confusion matrix: ", end="\n")
print(confusionMatrix)
print("we have multiple classified data = setosa, versicolor, virginica : 0,1,2 <- in that order")
print("accuracy: " + str(algorithm.Util.getAccuracy(confusionMatrix)))
print("macro precision: " + str(algorithm.Util.getMacroPrecision(confusionMatrix)))
print("macro recall: " + str(algorithm.Util.getMacroRecall(confusionMatrix)))
print("macro f1 score: " + str(algorithm.Util.getMacroF1Score(confusionMatrix)))

NOTICE: MODEL Iris LOADED SUCCESSFULLY!

Epoch 0 / 200
1/1 [==============================] loss:0
Epoch 1 / 200
1/1 [==============================] loss:0
Epoch 2 / 200
1/1 [==============================] loss:0
Epoch 3 / 200
1/1 [==============================] loss:0
Epoch 4 / 200
1/1 [==============================] loss:0
Epoch 5 / 200
1/1 [==============================] loss:0
Epoch 6 / 200
1/1 [==============================] loss:0
Epoch 7 / 200
1/1 [==============================] loss:0
Epoch 8 / 200
1/1 [==============================] loss:0
Epoch 9 / 200
1/1 [==============================] loss:0
Epoch 10 / 200
1/1 [==============================] loss:0
Epoch 11 / 200
1/1 [==============================] loss:0
Epoch 12 / 200
1/1 [==============================] loss:0
Epoch 13 / 200
1/1 [==============================] loss:0
Epoch 14 / 200
1/1 [==============================] loss:0
Epoch 15 / 200
1/1 [==============================] loss:0
Epoch 16 / 200
1/1 [=====

In [89]:
#config 4
#redefine model
Model_Test4 = FFNN('./file/models/Iris_Model_Test.json')
Model_Test4.setLearningRate(0.01)
#train using training data
Model_Test4.train(x_train4, y_train4, 200, 1, 0.0000000000001)
#predict validate
mymodel_predict = Model_Test4.batch_predict(x_validate4)
#reformat answers
finalAns = []
for item in mymodel_predict:
    if item[0]>0.5:
        finalAns.append(1)
    else:
        finalAns.append(0)
print(finalAns)
#reformat y_validate
y_validate = [y_validate4[i][0] for i in range (len(y_validate4))]
print(y_validate)
confusionMatrix = algorithm.Util.findConfusionMatrix(y_validate, finalAns)
print("confusion matrix: ", end="\n")
print(confusionMatrix)
print("we have multiple classified data = setosa, versicolor, virginica : 0,1,2 <- in that order")
print("accuracy: " + str(algorithm.Util.getAccuracy(confusionMatrix)))
print("macro precision: " + str(algorithm.Util.getMacroPrecision(confusionMatrix)))
print("macro recall: " + str(algorithm.Util.getMacroRecall(confusionMatrix)))
print("macro f1 score: " + str(algorithm.Util.getMacroF1Score(confusionMatrix)))

NOTICE: MODEL Iris LOADED SUCCESSFULLY!

Epoch 0 / 200
1/1 [==============================] loss:0
Epoch 1 / 200
1/1 [==============================] loss:0
Epoch 2 / 200
1/1 [==============================] loss:0
Epoch 3 / 200
1/1 [==============================] loss:0
Epoch 4 / 200
1/1 [==============================] loss:0
Epoch 5 / 200
1/1 [==============================] loss:0
Epoch 6 / 200
1/1 [==============================] loss:0
Epoch 7 / 200
1/1 [==============================] loss:0
Epoch 8 / 200
1/1 [==============================] loss:0
Epoch 9 / 200
1/1 [==============================] loss:0
Epoch 10 / 200
1/1 [==============================] loss:0
Epoch 11 / 200
1/1 [==============================] loss:0
Epoch 12 / 200
1/1 [==============================] loss:0
Epoch 13 / 200
1/1 [==============================] loss:0
Epoch 14 / 200
1/1 [==============================] loss:0
Epoch 15 / 200
1/1 [==============================] loss:0
Epoch 16 / 200
1/1 [=====

In [90]:
#config 5
#redefine model
Model_Test5 = FFNN('./file/models/Iris_Model_Test.json')
Model_Test5.setLearningRate(0.01)
#train using training data
Model_Test5.train(x_train5, y_train5, 200, 1, 0.0000000000001)
#predict validate
mymodel_predict = Model_Test5.batch_predict(x_validate5)
#reformat answers
finalAns = []
for item in mymodel_predict:
    if item[0]>0.5:
        finalAns.append(1)
    else:
        finalAns.append(0)
print(finalAns)
#reformat y_validate
y_validate = [y_validate5[i][0] for i in range (len(y_validate5))]
print(y_validate)
confusionMatrix = algorithm.Util.findConfusionMatrix(y_validate, finalAns)
print("confusion matrix: ", end="\n")
print(confusionMatrix)
print("we have multiple classified data = setosa, versicolor, virginica : 0,1,2 <- in that order")
print("accuracy: " + str(algorithm.Util.getAccuracy(confusionMatrix)))
print("macro precision: " + str(algorithm.Util.getMacroPrecision(confusionMatrix)))
print("macro recall: " + str(algorithm.Util.getMacroRecall(confusionMatrix)))
print("macro f1 score: " + str(algorithm.Util.getMacroF1Score(confusionMatrix)))

NOTICE: MODEL Iris LOADED SUCCESSFULLY!

Epoch 0 / 200
1/1 [==============================] loss:0
Epoch 1 / 200
1/1 [==============================] loss:0
Epoch 2 / 200
1/1 [==============================] loss:0
Epoch 3 / 200
1/1 [==============================] loss:0
Epoch 4 / 200
1/1 [==============================] loss:0
Epoch 5 / 200
1/1 [==============================] loss:0
Epoch 6 / 200
1/1 [==============================] loss:0
Epoch 7 / 200
1/1 [==============================] loss:0
Epoch 8 / 200
1/1 [==============================] loss:0
Epoch 9 / 200
1/1 [==============================] loss:0
Epoch 10 / 200
1/1 [==============================] loss:0
Epoch 11 / 200
1/1 [==============================] loss:0
Epoch 12 / 200
1/1 [==============================] loss:0
Epoch 13 / 200
1/1 [==============================] loss:0
Epoch 14 / 200
1/1 [==============================] loss:0
Epoch 15 / 200
1/1 [==============================] loss:0
Epoch 16 / 200
1/1 [=====

In [91]:
#config 6
#redefine model
Model_Test6 = FFNN('./file/models/Iris_Model_Test.json')
Model_Test6.setLearningRate(0.01)
#train using training data
Model_Test6.train(x_train6, y_train6, 200, 1, 0.0000000000001)
#predict validate
mymodel_predict = Model_Test6.batch_predict(x_validate6)
#reformat answers
finalAns = []
for item in mymodel_predict:
    if item[0]>0.5:
        finalAns.append(1)
    else:
        finalAns.append(0)
print(finalAns)
#reformat y_validate
y_validate = [y_validate6[i][0] for i in range (len(y_validate6))]
print(y_validate)
confusionMatrix = algorithm.Util.findConfusionMatrix(y_validate, finalAns)
print("confusion matrix: ", end="\n")
print(confusionMatrix)
print("we have multiple classified data = setosa, versicolor, virginica : 0,1,2 <- in that order")
print("accuracy: " + str(algorithm.Util.getAccuracy(confusionMatrix)))
print("macro precision: " + str(algorithm.Util.getMacroPrecision(confusionMatrix)))
print("macro recall: " + str(algorithm.Util.getMacroRecall(confusionMatrix)))
print("macro f1 score: " + str(algorithm.Util.getMacroF1Score(confusionMatrix)))

NOTICE: MODEL Iris LOADED SUCCESSFULLY!

Epoch 0 / 200
1/1 [==============================] loss:0
Epoch 1 / 200
1/1 [==============================] loss:0
Epoch 2 / 200
1/1 [==============================] loss:0
Epoch 3 / 200
1/1 [==============================] loss:0
Epoch 4 / 200
1/1 [==============================] loss:0
Epoch 5 / 200
1/1 [==============================] loss:0
Epoch 6 / 200
1/1 [==============================] loss:0
Epoch 7 / 200
1/1 [==============================] loss:0
Epoch 8 / 200
1/1 [==============================] loss:0
Epoch 9 / 200
1/1 [==============================] loss:0
Epoch 10 / 200
1/1 [==============================] loss:0
Epoch 11 / 200
1/1 [==============================] loss:0
Epoch 12 / 200
1/1 [==============================] loss:0
Epoch 13 / 200
1/1 [==============================] loss:0
Epoch 14 / 200
1/1 [==============================] loss:0
Epoch 15 / 200
1/1 [==============================] loss:0
Epoch 16 / 200
1/1 [=====

In [92]:
#config 7
#redefine model
Model_Test7 = FFNN('./file/models/Iris_Model_Test.json')
Model_Test7.setLearningRate(0.01)
# train using training data
# liat perbedaan model sebelum dan sesudah train
# print("Model sebelum train")
# print(Model_Test7)
Model_Test7.train(x_train7, y_train7, 200, 1, 0.0000000000001)
# print("Model setelah train")
# print(Model_Test7)
#predict validate
mymodel_predict = Model_Test7.batch_predict(x_validate7)
#reformat answers
finalAns = []
for item in mymodel_predict:
    if item[0]>0.5:
        finalAns.append(1)
    else:
        finalAns.append(0)
print(finalAns)
#reformat y_validate
y_validate = [y_validate7[i][0] for i in range (len(y_validate7))]
print(y_validate)
confusionMatrix = algorithm.Util.findConfusionMatrix(y_validate, finalAns)
print("confusion matrix: ", end="\n")
print(confusionMatrix)
print("we have multiple classified data = setosa, versicolor, virginica : 0,1,2 <- in that order")
print("accuracy: " + str(algorithm.Util.getAccuracy(confusionMatrix)))
print("macro precision: " + str(algorithm.Util.getMacroPrecision(confusionMatrix)))
print("macro recall: " + str(algorithm.Util.getMacroRecall(confusionMatrix)))
print("macro f1 score: " + str(algorithm.Util.getMacroF1Score(confusionMatrix)))

NOTICE: MODEL Iris LOADED SUCCESSFULLY!

Epoch 0 / 200
1/1 [==============================] loss:0
Epoch 1 / 200
1/1 [==============================] loss:0
Epoch 2 / 200
1/1 [==============================] loss:0
Epoch 3 / 200
1/1 [==============================] loss:0
Epoch 4 / 200
1/1 [==============================] loss:0
Epoch 5 / 200
1/1 [==============================] loss:0
Epoch 6 / 200
1/1 [==============================] loss:0
Epoch 7 / 200
1/1 [==============================] loss:0
Epoch 8 / 200
1/1 [==============================] loss:0
Epoch 9 / 200
1/1 [==============================] loss:0
Epoch 10 / 200
1/1 [==============================] loss:0
Epoch 11 / 200
1/1 [==============================] loss:0
Epoch 12 / 200
1/1 [==============================] loss:0
Epoch 13 / 200
1/1 [==============================] loss:0
Epoch 14 / 200
1/1 [==============================] loss:0
Epoch 15 / 200
1/1 [==============================] loss:0
Epoch 16 / 200
1/1 [=====

In [93]:
#config 8
#redefine model
Model_Test8 = FFNN('./file/models/Iris_Model_Test.json')
Model_Test8.setLearningRate(0.01)
#train using training data
Model_Test8.train(x_train8, y_train8, 200, 1, 0.0000000000001)
#predict validate
mymodel_predict = Model_Test8.batch_predict(x_validate8)
#reformat answers
finalAns = []
for item in mymodel_predict:
    if item[0]>0.5:
        finalAns.append(1)
    else:
        finalAns.append(0)
print(finalAns)
#reformat y_validate
y_validate = [y_validate8[i][0] for i in range (len(y_validate8))]
print(y_validate)
confusionMatrix = algorithm.Util.findConfusionMatrix(y_validate, finalAns)
print("confusion matrix: ", end="\n")
print(confusionMatrix)
print("we have multiple classified data = setosa, versicolor, virginica : 0,1,2 <- in that order")
print("accuracy: " + str(algorithm.Util.getAccuracy(confusionMatrix)))
print("macro precision: " + str(algorithm.Util.getMacroPrecision(confusionMatrix)))
print("macro recall: " + str(algorithm.Util.getMacroRecall(confusionMatrix)))
print("macro f1 score: " + str(algorithm.Util.getMacroF1Score(confusionMatrix)))

NOTICE: MODEL Iris LOADED SUCCESSFULLY!

Epoch 0 / 200
1/1 [==============================] loss:0
Epoch 1 / 200
1/1 [==============================] loss:0
Epoch 2 / 200
1/1 [==============================] loss:0
Epoch 3 / 200
1/1 [==============================] loss:0
Epoch 4 / 200
1/1 [==============================] loss:0
Epoch 5 / 200
1/1 [==============================] loss:0
Epoch 6 / 200
1/1 [==============================] loss:0
Epoch 7 / 200
1/1 [==============================] loss:0
Epoch 8 / 200
1/1 [==============================] loss:0
Epoch 9 / 200
1/1 [==============================] loss:0
Epoch 10 / 200
1/1 [==============================] loss:0
Epoch 11 / 200
1/1 [==============================] loss:0
Epoch 12 / 200
1/1 [==============================] loss:0
Epoch 13 / 200
1/1 [==============================] loss:0
Epoch 14 / 200
1/1 [==============================] loss:0
Epoch 15 / 200
1/1 [==============================] loss:0
Epoch 16 / 200
1/1 [=====

In [94]:
#config 9
#redefine model
Model_Test9 = FFNN('./file/models/Iris_Model_Test.json')
Model_Test9.setLearningRate(0.01)
#train using training data
Model_Test9.train(x_train9, y_train9, 200, 1, 0.0000000000001)
#predict validate
mymodel_predict = Model_Test9.batch_predict(x_validate9)
#reformat answers
finalAns = []
for item in mymodel_predict:
    if item[0]>0.5:
        finalAns.append(1)
    else:
        finalAns.append(0)
print(finalAns)
#reformat y_validate
y_validate = [y_validate9[i][0] for i in range (len(y_validate9))]
print(y_validate)
confusionMatrix = algorithm.Util.findConfusionMatrix(y_validate, finalAns)
print("confusion matrix: ", end="\n")
print(confusionMatrix)
print("we have multiple classified data = setosa, versicolor, virginica : 0,1,2 <- in that order")
print("accuracy: " + str(algorithm.Util.getAccuracy(confusionMatrix)))
print("macro precision: " + str(algorithm.Util.getMacroPrecision(confusionMatrix)))
print("macro recall: " + str(algorithm.Util.getMacroRecall(confusionMatrix)))
print("macro f1 score: " + str(algorithm.Util.getMacroF1Score(confusionMatrix)))

NOTICE: MODEL Iris LOADED SUCCESSFULLY!

Epoch 0 / 200
1/1 [==============================] loss:0
Epoch 1 / 200
1/1 [==============================] loss:0
Epoch 2 / 200
1/1 [==============================] loss:0
Epoch 3 / 200
1/1 [==============================] loss:0
Epoch 4 / 200
1/1 [==============================] loss:0
Epoch 5 / 200
1/1 [==============================] loss:0
Epoch 6 / 200
1/1 [==============================] loss:0
Epoch 7 / 200
1/1 [==============================] loss:0
Epoch 8 / 200
1/1 [==============================] loss:0
Epoch 9 / 200
1/1 [==============================] loss:0
Epoch 10 / 200
1/1 [==============================] loss:0
Epoch 11 / 200
1/1 [==============================] loss:0
Epoch 12 / 200
1/1 [==============================] loss:0
Epoch 13 / 200
1/1 [==============================] loss:0
Epoch 14 / 200
1/1 [==============================] loss:0
Epoch 15 / 200
1/1 [==============================] loss:0
Epoch 16 / 200
1/1 [=====

In [95]:
# for this test, configuration 5,6,7, and 7 ties in performance
# we pick configuration 7's model to predict with the test data

mymodel_predict = Model_Test7.batch_predict(x_test)
#reformat answers
finalAns = []
for item in mymodel_predict:
    if item[0]>0.5:
        finalAns.append(1)
    else:
        finalAns.append(0)
print(finalAns)
#reformat y_validate
y_final = [y_test[i][0] for i in range (len(y_test))]
print(y_final)
confusionMatrix = algorithm.Util.findConfusionMatrix(y_final, finalAns)
print("confusion matrix: ", end="\n")
print(confusionMatrix)
print("we have multiple classified data = setosa, versicolor, virginica : 0,1,2 <- in that order")
print("accuracy: " + str(algorithm.Util.getAccuracy(confusionMatrix)))
print("macro precision: " + str(algorithm.Util.getMacroPrecision(confusionMatrix)))
print("macro recall: " + str(algorithm.Util.getMacroRecall(confusionMatrix)))
print("macro f1 score: " + str(algorithm.Util.getMacroF1Score(confusionMatrix)))

[1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]
[0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2]
confusion matrix: 
[[0, 2, 0], [0, 50, 0], [0, 50, 0]]
we have multiple classified data = setosa, versicolor, virginica : 0,1,2 <- in that order
accuracy: 0.49019607843137253
macro precision: 0.16339869281045752
macro recall: 0.3333333333333333
macro f1 score: 0.21929824561403508


In [96]:
#save model 7 
save(Model_Test7, model_name="Model_Test7", file_path='./file/models/Model_Test7.json')

In [105]:
newinstances = [
    [1.5, 2.4 , 1.2, 0.9],
    [2.6, 3.3, 2.2, 1.9],
    [1.7, 2.3, 0.2, 2.9],
    [3.8, 2.4, 1.2, 0.7],
    [2.7, 2.4, 1.1, 2.7],
    [3.8, 3.6, 2.6, 1.9],
    [1.9, 2.8, 3.2, 2.3],
    [2.9, 2.2, 2.4, 0.1],
    [3.9, 1.4, 1.7, 0.2],
    [4.9, 1.6, 0.7, 0.3],
    ]
print("New instances:",newinstances)
res = Model_Test7.batch_predict(newinstances)
final = []
for item in res:
    if item[0]>0.5:
        final.append(1)
    else:
        final.append(0)
print("New instances result:",final)

New instances: [[1.5, 2.4, 1.2, 0.9], [2.6, 3.3, 2.2, 1.9], [1.7, 2.3, 0.2, 2.9], [3.8, 2.4, 1.2, 0.7], [2.7, 2.4, 1.1, 2.7], [3.8, 3.6, 2.6, 1.9], [1.9, 2.8, 3.2, 2.3], [2.9, 2.2, 2.4, 0.1], [3.9, 1.4, 1.7, 0.2], [4.9, 1.6, 0.7, 0.3]]
New instances result: [1, 1, 1, 1, 1, 1, 1, 1, 1, 1]


# Confusion matrix and sklearn analysis

Dari sklearn, dapat dilihat bahwa confusion matrix menghasilkan

[[50, 0, 0], 

[0, 28, 22],
 
[0, 0, 50]]

Matrix menunjukkan bahwa sklearn berhasil mengklasifikasi 50 setosa sebagai setosa, 28 versicolor sebagai versicolor, dan 50 virginica sebagai virginica. Adapun kesalahan klasifikasi, yaitu prediksi 22 versicolor sebagai virginica.

Akurasi, presisi, recall, dan f1 didapat sebagai berikut:

accuracy: 0.8533333333333334

macro precision: 0.8981481481481483

macro recall: 0.8533333333333334

macro f1 score: 0.8751674067808557

Akurasi dihitung dengan rasio jumlah true negative dan true positif dengan total, yaitu 128/150 atau 0.8533... sehingga didapat perhitungan akurasi benar.

Presisi dihitung dengan rasio true positive dengan total true positive dan false positive, yaitu (50/50 + 28/28 + 50/72)/3 atau 0.89814... sehingga didapat perhitungan akurasi benar.

Recall dihitung dengan rasio true positive dengan true positive + false negative, yaitu (50/50 + 28/50 + 50/50)/3 atau 0.8533... sehingga didapat perhitungan akurasi benar.

F1 dihitung dengan rasio perkalian presisi dan recall dengan penjumlahan presisi dan recall, yaitu 2* (0.89814... * 0.8533...) / (0.89814... + 0.8533...) atau 0.8533... sehingga didapat perhitungan akurasi benar.

Akurasi berarti berapa banyak klasifikasi data yang benar. Dengan akurasi 0.8533..., berarti model mengklasifikasi 85.33..% data dengan benar dan nilai yang tinggi berarti model akan memiliki peluang klasifikasi benar yang tinggi.

Presisi menentukan berapa banyak true positive jika dibandingkan dengan true positive + negatif yang ternyata positif sehingga nilai tinggi lebih baik. Begitupun dengan recall yang dibandingkan dengan true positive + positive yang ternyata negatif. Kedua nilai ini yang tinggi akan menentukan kualitas model lebih baik dibanding akurasi karena akurasi tidak memperhitungkan FP dan FN. Perhitungan kedua nilai dihitung lewat nilai F1.

Kemudian jika dibandingkan dengan hasil dari model buatan

confusion matrix: 

[[0, 50, 0], [0, 50, 0], [0, 50, 0]]

we have multiple classified data = setosa, versicolor, virginica : 0,1,2 <- in that order

accuracy: 0.3333333333333333

macro precision: 0.1111111111111111

macro recall: 0.3333333333333333

macro f1 score: 0.16666666666666666

akurasi yang didapat cukup rendah dengan nilai 0.33... dengan presisi yang lebih rendah 0.111... namun recall lebih tinggi sebesar 0.333. Hasil F1 didapat adalah 0.167. Hal ini terjadi karena banyak klasifikasi yang masuk ke versicolor yang banyak menyebabkan false positive dan false negative yang mengurangi kualitas model.

# Split 9:1 training Analysis

Hasil training split data adalah sebagai berikut

[0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2]

confusion matrix: 

[[0, 2, 0], [0, 50, 0], [0, 50, 0]]

we have multiple classified data = setosa, versicolor, virginica : 0,1,2 <- in that order

accuracy: 0.49019607843137253

macro precision: 0.16339869281045752

macro recall: 0.3333333333333333

macro f1 score: 0.21929824561403508

Setelah dilakukan training, terlihat ada kenaikan kualitas model. Akurasi naik dari 0.33... menjadi 0.49 yang berarti klasifikasi TP dan TN akan semakin akurat. Presisi mengalami kenaikan juga dari 0.11... menjadi 0.163 yang berarti jumlah negatif yang ternyata positif berkurang. Recall tetap sehingga hasil training tidak memengaruhi jumlah false negative. Karena salah satu dari presisi dan recall mengalami kenaikan, F1 juga mengalami kenaikan menjadi 0.21. 

Kesimpulannya adalah setelah training 9:1, model mengalami kenaikan kualitas.